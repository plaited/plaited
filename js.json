{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "plaited",
    "paths": {
      "plaited/jsx-runtime": [
        "./src/jsx/runtime.ts"
      ],
      "plaited/jsx-dev-runtime": [
        "./src/jsx/dev-runtime.ts"
      ]
    }
  }
}