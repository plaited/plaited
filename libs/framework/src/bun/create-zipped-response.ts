export const jsMimeTypes = ['text/javascript;charset=utf-8', 'application/json;charset=utf-8'] as const

export const isJsMimeType = (type: string): type is (typeof jsMimeTypes)[number] =>
  jsMimeTypes.includes(type as (typeof jsMimeTypes)[number])

export const createZippedResponse = (str: string, type: (typeof jsMimeTypes)[number]) => {
  const compressed = Bun.gzipSync(Buffer.from(str))
  return new Response(compressed, {
    headers: {
      'content-type': type,
      'content-encoding': 'gzip',
    },
  })
}
