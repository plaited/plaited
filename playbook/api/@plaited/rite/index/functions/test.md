**@plaited/rite** ( [Readme](../../README.md) \| API )

***

[Plaited Typedocs](../../../../modules.md) / [@plaited/rite](../../modules.md) / [index](../README.md) / test

# Function: test()

> **test**(`name`, `cb`): `void`

## Parameters

▪ **name**: `string`

▪ **cb**: `TestCallback`

## Returns

`void`

## Source

[libs/rite/src/test.ts:10](https://github.com/plaited/plaited/blob/0d4801d/libs/rite/src/test.ts#L10)

***

Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
