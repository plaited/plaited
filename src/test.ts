export * from './test/assert.types.js'
export * from './test/assert.constants.js'
export { assert } from './test/assert.js'
export { findByAttribute } from './test/find-by-attribute.js'
export { findByText } from './test/find-by-text.js'
export { fireEvent } from './test/fire-event.js'
export { match } from './test/match.js'
export { throws } from './test/throws.js'
export { wait } from './utils/wait.js'
export { AssertionError, MissingTestParamsError, TimeoutError } from './test/errors.js'
export { usePlay } from './test/use-play.js'
