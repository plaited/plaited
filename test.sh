#!/bin/bash

# Run yarn
yarn install --immutable

# Run yarn build
yarn build

# Run test
yarn test
