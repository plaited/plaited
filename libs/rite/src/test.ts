/// <reference lib="dom" />
/// <reference lib="dom.iterable" />

import { TestResultError, TestResult, BrowserSessionResult } from '@web/test-runner-core/browser/session.js'
import { assert, Assertion, AssertionError } from './assert.js'

type TestCallback = (t: Assertion) => Promise<void> | void

interface Test {
  (name: string, cb: TestCallback): void
  skip: (name: string, cb: TestCallback) => void
}

type TestCase = () => Promise<boolean> | boolean

class TestRunner {
  #completedTests = 0
  #failedTests = 0
  #errors: TestResultError[] = []
  #cases: TestCase[] = []
  #timeout = 5_000
  #tests: TestResult[] = []
  updateTimeout(time: number) {
    this.#timeout = time
  }
  get timeout() {
    return this.#timeout
  }
  addCase(test: TestCase) {
    this.#cases.push(test)
  }
  addResult(result: TestResult) {
    this.#tests.push(result)
  }
  addError(obj: TestResultError) {
    this.#errors.push(obj)
  }
  async run(name: string) {
    for (const testCase of this.#cases) {
      try {
        const passed = await testCase()
        this.#completedTests++
        if (!passed) {
          this.#failedTests++
        }
      } catch (error) {
        console.error(error)
      }
    }

    if (this.#completedTests === this.#cases.length) {
      const results: BrowserSessionResult = {
        passed: !this.#failedTests,
        testResults: {
          name,
          suites: [],
          tests: this.#tests,
        },
        errors: this.#errors,
      }
      return results
    } else {
      throw new AssertionError(`Not all test completed`)
    }
  }
}

declare global {
  interface Window {
    __rite_test_runner?: TestRunner
  }
}

window.__rite_test_runner = new TestRunner()

const throwTimeoutError = async () => {
  const timeout = window.__rite_test_runner.timeout
  await assert.wait(timeout)
  throw new Error(`test takes longer than ${timeout}ms to complete`)
}

export const test: Test = (name, callback) => {
  window.__rite_test_runner.addCase(async () => {
    const startTime = performance.now()
    const result: TestResult = {
      name,
      passed: true,
      skipped: false,
    }
    try {
      await Promise.race([callback(assert), throwTimeoutError()])
      const msg = `✓ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
    } catch (error) {
      const msg = `✗ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
      result.passed = false
      if (error instanceof AssertionError) {
        const obj: Pick<TestResultError, 'actual' | 'expected' | 'message'> = JSON.parse(error.message)
        result.error = { name, stack: `${error.stack}`, ...obj }
        window.__rite_test_runner.addError({ name, ...obj })
      } else {
        throw new Error(error)
      }
    }
    result.duration = performance.now() - startTime
    window.__rite_test_runner.addResult(result)
    return result.passed
  })
}
test.skip = (name: string, _: TestCallback) => {
  window.__rite_test_runner.addCase(() => {
    const result = {
      name,
      passed: true,
      skipped: true,
    }
    window.__rite_test_runner.addResult(result)
    return true
  })
}
