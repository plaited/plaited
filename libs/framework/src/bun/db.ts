import { Database } from "bun:sqlite";

export const db = new Database(":memory:");
db.query(`create table templates;`);