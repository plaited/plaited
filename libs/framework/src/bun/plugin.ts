import type {  BunPlugin } from "bun";
import { db } from "./db.js";

export const plugin: BunPlugin = {
  name: "Plaited loader",
  setup(build) {
    build.onResolve({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // mark external all non-template files
    })
    build.onLoad({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // modify the contents of the file to be an export object.
      return {
        contents,
        loader: "tsx",
      };
    });
  },
}