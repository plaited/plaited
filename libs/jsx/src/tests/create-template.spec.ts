import { test, expect } from 'bun:test'
import { createTemplate as h, css, Fragment, FT, Template } from '../index.js'
import beautify from 'beautify'

const render = (tpl: Template) => {
  const template = document.createElement('template')
  template.content.append(tpl.content)
  return beautify(template.innerHTML, { format: 'html' })
}

test('createTemplate: Self closing - html', () => {
  expect(render(h('input', { type: 'text' }))).toMatchSnapshot()
})

test('createTemplate: Self closing - svg', () => {
  expect(render(h('polygon', { points: '0,100 50,25 50,75 100,0' }))).toMatchSnapshot()
})

test('createTemplate: Falsey - undefined', () => {
  expect(render(h('div', { children: undefined }))).toMatchSnapshot()
})

test('createTemplate: Falsey - null', () => {
  expect(render(h('div', { children: null }))).toMatchSnapshot()
})

test('createTemplate: Falsey - false', () => {
  expect(render(h('div', { children: false }))).toMatchSnapshot()
})

test('createTemplate: Falsey - ""', () => {
  expect(render(h('div', { children: '' }))).toMatchSnapshot()
})

test('createTemplate: Falsey - 0', () => {
  expect(render(h('div', { children: 0 }))).toMatchSnapshot()
})

test('createTemplate: Falsey - NaN', () => {
  expect(render(h('div', { children: NaN }))).toMatchSnapshot()
})

test('createTemplate: Bad template - NaN', () => {
  expect(render(h('div', { children: { string: 'string' } }))).toMatchSnapshot()
})

test('createTemplate: Conditional', () => {
  expect(render(h('div', { children: true && 'hello' }))).toMatchSnapshot()
})

test('createTemplate: Style attribute', () => {
  const { content } = h('div', {
    style: { backgroundColor: 'blue', margin: '12px', '--cssVar': 'red' },
    children: 'styles',
  })
  const style = (content as HTMLDivElement).style.cssText
  expect(style).toMatchSnapshot()
})

test('createTemplate: data-trigger attribute', () =>
  expect(
    render(
      h('div', {
        'data-trigger': {
          click: 'random',
          focus: 'thing',
        },
        children: 'triggers',
      }),
    ),
  ).toMatchSnapshot())

test('createTemplate: kebab-case attributes', () => {
  expect(
    render(
      h('div', {
        dataTrigger: { click: 'click' },
        dataAddress: 'address',
        dataMode: 'mode',
        xPrefix: 'x',
      }),
    ),
  ).toMatchSnapshot()
})

test('createTemplate: Array of templates', () =>
  expect(
    render(
      h('ul', {
        children: Array.from(Array(10).keys()).map((n) => h('li', { children: `${n}` })),
      }),
    ),
  ).toMatchSnapshot())

test('createTemplate: Should throw with attribute starting with on', () => {
  expect(() => {
    h('div', {
      children: h('template', {
        shadowrootmode: 'closed',
        children: h('img', {
          src: 'nonexistent.png',
          onerror: "alert('xss!')",
        }),
      }),
    })
  }).toThrow()
})

test('createTemplate: should throw on script tag', () => {
  expect(() => {
    h('script', { type: 'module', src: 'main.js' })
  }).toThrow()
})

test('createTemplate: Should not throw on script tag with trusted attribute', () => {
  expect(render(h('script', { type: 'module', src: 'main.js', trusted: true }))).toMatchSnapshot()
})

test('createTemplate: Escapes children', () => {
  const scriptContent = `<script type="text/javascript">
const hostRegex = /^https?://([^/]+)/.*$/i;
const host = document.URL.replace(hostRegex, '$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script>`
  const { content } = h('div', { children: scriptContent })
  expect(content.firstElementChild).toMatchSnapshot()
})

test('createTemplate: Does not escape children when trusted', () => {
  const scriptContent = `<script type="text/javascript">
const hostRegex = /^https?://([^/]+)/.*$/i;
const host = document.URL.replace(hostRegex, '$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script>`
  const { content } = h('div', { trusted: true, children: scriptContent })
  expect(content.firstElementChild.outerHTML).toMatchSnapshot()
})

const Template: FT = (attrs) => h('template', attrs)

test('createTemplate: Non declarative shadow DOM template', () => {
  const List: FT = ({ children }) =>
    h('ul', {
      children: [
        Template({ children: h('span', { children: 'I am a span!!!' }) }),
        ...(Array.isArray(children) ? children : [children]),
      ],
    })

  expect(
    render(
      h(List, {
        children: Array.from(Array(10).keys()).map((n) => h('li', { children: `item-${n}` })),
      }),
    ),
  ).toMatchSnapshot()
})

test('createTemplate: Fragment', () => {
  expect(
    render(
      h(Fragment, {
        children: Array.from(Array(6).keys())
          .reverse()
          .map((n) => h('li', { children: n > 0 ? `In ${n}` : 'Blast Off!!!' })),
      }),
    ),
  ).toMatchSnapshot()
})

const nestedDeclarativeStyles = css`
  .nested-label {
    font-weight: bold;
  }
`

const NestedCustomElement: FT = ({ children, stylesheet }) =>
  h('nested-component', {
    stylesheet,
    children: [
      Template({
        shadowrootmode: 'open',
        shadowrootdelegatesfocus: true,
        children: [
          h('span', {
            className: nestedDeclarativeStyles[0]['nested-label'],
            ...nestedDeclarativeStyles[1],
            children: 'inside nested template',
          }),
          h('slot', { name: 'nested' }),
        ],
      }),
      ...(Array.isArray(children) ? children : [children]),
    ],
  })

test('createTemplate: CustomElement hoisting its styles', () => {
  const el = h(NestedCustomElement, {})
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const nestedHostStyles = css`
  host: {
    display: flex;
    flex-direction: column;
  }
`

test('createTemplate: CustomElement with declarative shadow dom & host styles', () => {
  const el = h(NestedCustomElement, { ...nestedHostStyles[1] })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const nestedChildrenStyles = css`
  .slotted-paragraph {
    color: rebeccapurple;
  }
`

test('createTemplate: CustomElement with styled slotted component', () => {
  const el = h(NestedCustomElement, {
    children: [
      h('p', {
        slot: 'nested',
        className: nestedChildrenStyles[0]['slotted-paragraph'],
        ...nestedChildrenStyles[1],
        children: 'slotted paragraph',
      }),
    ],
  })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const TopCustomElement: FT = ({ children, stylesheet }) =>
  h('top-component', {
    stylesheet,
    children: [
      Template({
        shadowrootdelegatesfocus: true,
        shadowrootmode: 'open',
        children: h(NestedCustomElement, {
          children: h('p', {
            slot: 'nested',
            className: nestedChildrenStyles[0]['slotted-paragraph'],
            ...nestedChildrenStyles[1],
            children: 'slotted paragraph',
          }),
        }),
      }),
      ...(Array.isArray(children) ? children : [children]),
    ],
  })

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom', () => {
  const el = h(TopCustomElement, {})
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const hostStyles = css`
  :host {
    display: block;
  }
`

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles', () => {
  const el = h(TopCustomElement, { ...hostStyles[1] })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const imageStyles = css`
  .image {
    width: 100%;
    aspect-ratio: 16 / 9;
  }
`

test('createTemplate: CustomElement with declarative shadow dom and nested declarative shadow dom plus host styles and child', () => {
  const el = h(TopCustomElement, {
    ...hostStyles[1],
    children: h('img', { className: imageStyles[0].image, ...imageStyles[1] }),
  })
  expect({ content: render(el), stylesheets: el.stylesheets }).toMatchSnapshot()
})

const sheet1 = css`
  .a {
    width: 100%;
  }
`
const sheet2 = css`
  .a {
    width: 100%;
  }
`

const sheet3 = css`
  .a {
    color: blue;
  }
`

test('ssr: Properly hoist and deduplicates multiple stylesheets on a single node', () => {
  expect(
    h('div', {
      stylesheet: [sheet1[1].stylesheet, sheet2[1].stylesheet, sheet3[1].stylesheet],
    }).stylesheets.size,
  ).toBe(2)
})

test('createTemplate: Trims whitespace', () => {
  expect(
    render(
      h('div', {
        children: '   trims white-space    ',
      }),
    ),
  ).toMatchSnapshot()
})
