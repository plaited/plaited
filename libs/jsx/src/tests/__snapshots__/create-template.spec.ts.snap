// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createTemplate: self closing - html 1`] = `
{
  "content": "<input type="text"/>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: self closing - svg 1`] = `
{
  "content": "<polygon points="0,100 50,25 50,75 100,0"/>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: falsey - undefined 1`] = `
{
  "content": "<div></div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: falsey - null 1`] = `
{
  "content": "<div></div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: falsey - false 1`] = `
{
  "content": "<div></div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: falsey - "" 1`] = `
{
  "content": "<div></div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: falsey - 0 1`] = `
{
  "content": "<div></div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: falsey - NaN 1`] = `
{
  "content": "<div></div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: conditional 1`] = `
{
  "content": "<div>hello</div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: style attribute 1`] = `
{
  "content": "<div style="background-color:blue; margin:12px;">styles</div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: data-trigger attribute 1`] = `
{
  "content": "<div data-trigger="click->random focus->thing">triggers</div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: array of PlaitedElements 1`] = `
{
  "content": "<div><li>0</li><li>1</li><li>2</li><li>3</li><li>4</li><li>5</li><li>6</li><li>7</li><li>8</li><li>9</li></div>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: should not throw on script tag with trusted attribute 1`] = `
{
  "content": "<script type="module" src="main.js"></script>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: escapes children 1`] = `
{
  "content": 
"<div>&lt;script type=&quot;text/javascript&quot;&gt;
const hostRegex = /^https?://([^/]+)/.*$/i;
const host = document.URL.replace(hostRegex, &#39;$1&#39;);
const socket = new WebSocket(/);
const reload = () =&gt;{
  location.reload();
  console.log(&#39;...reloading&#39;);
};
socket.addEventListener(&#39;message&#39;, reload);
console.log(&#39;[plaited] listening for file changes&#39;);
&lt;/script&gt;</div>"
,
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: does not escape children when trusted 1`] = `
{
  "content": 
"<div><script type="text/javascript">
const hostRegex = /^https?://([^/]+)/.*$/i;
const host = document.URL.replace(hostRegex, '$1');
const socket = new WebSocket(/);
const reload = () =>{
  location.reload();
  console.log('...reloading');
};
socket.addEventListener('message', reload);
console.log('[plaited] listening for file changes');
</script></div>"
,
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: with slotted PlaitedElements 1`] = `
{
  "content": "<c-el><template shadowrootmode="open" shadowrootdelegatesfocus="true"><slot name="slot"></slot></template><li slot="slot">slot-0</li><li slot="slot">slot-1</li><li slot="slot">slot-2</li><li slot="slot">slot-3</li><li slot="slot">slot-4</li><li slot="slot">slot-5</li><li slot="slot">slot-6</li><li slot="slot">slot-7</li><li slot="slot">slot-8</li><li slot="slot">slot-9</li></c-el>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: Fragment PlaitedElements 1`] = `
{
  "content": "<li>item-0</li><li>item-1</li><li>item-2</li><li>item-3</li><li>item-4</li><li>item-5</li><li>item-6</li><li>item-7</li><li>item-8</li><li>item-9</li>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: custom element with child hoisting its styles 1`] = `
{
  "content": "<nested-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>.nested-label_LTU0Mz { font-weight: bold; }</style><span class="nested-label_LTU0Mz">inside nested template</span><slot name="nested"></slot></template></nested-component>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: custom element with child and host styles 1`] = `
{
  "content": "<nested-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>host: { display: flex; flex-direction: column; }.nested-label_LTU0Mz { font-weight: bold; }</style><span class="nested-label_LTU0Mz">inside nested template</span><slot name="nested"></slot></template></nested-component>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: custom element with styled slotted component 1`] = `
{
  "content": "<nested-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>.nested-label_LTU0Mz { font-weight: bold; }</style><span class="nested-label_LTU0Mz">inside nested template</span><slot name="nested"></slot></template><p class="slotted-paragraph_LTc5ND" slot="nested">slotted paragraph</p></nested-component>",
  "stylesheets":   
Set {
    ".slotted-paragraph_LTc5ND { color: rebeccapurple; }",
  }
,
}
`;

exports[`createTemplate: custom element with styles nested in custom element 1`] = `
{
  "content": "<top-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>.slotted-paragraph_LTc5ND { color: rebeccapurple; }</style><nested-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>.nested-label_LTU0Mz { font-weight: bold; }</style><span class="nested-label_LTU0Mz">inside nested template</span><slot name="nested"></slot></template><p class="slotted-paragraph_LTc5ND" slot="nested">slotted paragraph</p></nested-component></template></top-component>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: custom element with styles nested in custom element with styles 1`] = `
{
  "content": "<top-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>:host { display: block; }.slotted-paragraph_LTc5ND { color: rebeccapurple; }</style><nested-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>.nested-label_LTU0Mz { font-weight: bold; }</style><span class="nested-label_LTU0Mz">inside nested template</span><slot name="nested"></slot></template><p class="slotted-paragraph_LTc5ND" slot="nested">slotted paragraph</p></nested-component></template></top-component>",
  "stylesheets":   Set {},
}
`;

exports[`createTemplate: custom element with nested custom element and styled slotted element 1`] = `
{
  "content": "<top-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>:host { display: block; }.slotted-paragraph_LTc5ND { color: rebeccapurple; }</style><nested-component><template shadowrootmode="open" shadowrootdelegatesfocus="true"><style>.nested-label_LTU0Mz { font-weight: bold; }</style><span class="nested-label_LTU0Mz">inside nested template</span><slot name="nested"></slot></template><p class="slotted-paragraph_LTc5ND" slot="nested">slotted paragraph</p></nested-component></template><img class="image_Mjc4MD"/></top-component>",
  "stylesheets":   
Set {
    ".image_Mjc4MD { width: 100%; aspect-ratio: 16 / 9; }",
  }
,
}
`;

exports[`createTemplate: properly hoist and deduplicates multiple stylesheets on a single node 1`] = `
{
  "content": "<div></div>",
  "stylesheets":   
Set {
    ".a_NTA5Nj { width: 100%; }",
    ".a_LTIzMT { color: blue; }",
  }
,
}
`;
